### DISCUSSION TITLE
Can `TableView` host `Views` in cells?

### DISCUSSION THREAD

**Discussion by tig** [tui-cs/Terminal.Gui#2996] [Q&A]
@tznind did you ever contemplate having the content for a cell in `TableView` be an arbitrary `View`?

**Comment by tznind** [comment]
### Table View Limitations

`TableView` draws its content from a given row/column offset.  It takes the next x rows where x is the screen height and measures the contents.  Then it decides how many columns it can fit based on the size (width) of data in these rows and renders.

This means
- Rows can only ever be 1 in height
- As you scroll down columns can expand/contract as wider values are seen
- Scrolling increment is always done 1 row/column at a time

This was done to offer the best presentation and scaleability of tabular data.

---------------------

### Faking it

It is possible to achieve a similar behaviour to what you ask by holding a `View` in memory but not mounted and piping events to it.  This is now `TreeTableSource<T>` works. 

But notice how each 'row' of the table is still a line of the `TreeView<T>`  there are no arbitrary height elements.

Similarly `CheckBoxTableSourceWrapperBase` lets checkboxes/radio appear in the table although in this case it is just a visual effect.

---------------------
### A table of views
To layout a collection of `View` as a table, I think we would want a new purpose built view or introduce the concept of a layout manager.

For `TableLayout` you would want to inspect all the `View` being presented in all 'boxes' and then fiddle the layouts e.g.

- Ensure all View on a row have the same height (that of the max height view in the row).
- Allow free scrolling (i.e. be able to scroll 1 console cell across at a time rather than jumping table cell by table cell

![18](https://github.com/gui-cs/Terminal.Gui/assets/31306100/1248efc1-f78a-446a-a93f-15970f39918c)

Here is the CSS specification for how to layout a table of arbitrary sized 'stuff':
https://drafts.csswg.org/css2/#tables

See in particular the most exciting bit: [17.5.2. Table width algorithms: the table-layout property](https://drafts.csswg.org/css2/#width-layout)


---------------------

### Related concepts

I notice how most users expect `ScrollView` to auto resize based on what gets added to it.  Instead user has to manually measure their views and configure a set `ContentSize`.  Once we are in the buisness of measuring views and adjusting layouts dynamically we could probably accomodate this.

There is also opportunities for things like `FlowLayout` or `ColumnLayout` which would be nice to have.

**Comment by tig** [comment]
Excellent response. Thanks!

No real need for this now, I just was curious. I do like the idea of a `TableLayout`.
